### PR TITLE
Include a directory with Flutter assets

### PR DESCRIPTION
**Pull request by szakarias**
The `flutter_assets` directory is meant to eventually replace `app.flx`.

This PR also changes the way an asset is retrieved. On asset channel queries, engine now returns the path to the asset instead of returning the asset as byte data.

Before merging this I will have another PR ready which can upgrade the `project.pbxproj` file of existing projects to include `flutter_assets`.

Engine PR https://github.com/flutter/engine/pull/4343

### CODE REVIEW

File: packages/flutter/lib/src/services/asset_bundle.dart
Comment (Hixie) [inline packages/flutter/lib/src/services/asset_bundle.dart]: just `assert(bytes is Uint8List)`. If it's ever NOT a Uint8List this is a serious performance concern.

**Comment by szakarias** [comment]
cc @tvolkert

**Comment by tvolkert** [comment]
> Before merging this I will have another PR ready which can upgrade
> the project.pbxproj file of existing projects to include flutter_assets.

FYI @xster who's been working on a "flutter project upgrade" mechanism.

**Review by tvolkert — COMMENTED** [review]
Once the engine changes land, this should include the engine roll to the appropriate revision.

**Comment by xster** [comment]
I'm probably still at least a week away from a functioning flutter project upgrade mechanism. How does this PR work with flutter projects created in the past?

File: packages/flutter_tools/bin/xcode_backend.sh
Comment (szakarias) [reply packages/flutter_tools/bin/xcode_backend.sh]: Nope, I am however considering renaming the option to `assets-dir` when doing the FLX removal cleanup.

**Comment by szakarias** [comment]
Old flutter projects will need to get their `project.pbxproj` updated for this to work.  The PR for it is here #13011 .

File: packages/flutter_tools/bin/xcode_backend.sh
Comment (jakobr-google) [reply packages/flutter_tools/bin/xcode_backend.sh]: Rename sgtm. And the reason we're putting this in `derived_dir` rather than `build_dir` is that we need to reference it from `project.pbxproj` (like the current `app.flx`, `App.framework`, etc.), right?

**Review by jakobr-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/bin/xcode_backend.sh
Comment (szakarias) [reply packages/flutter_tools/bin/xcode_backend.sh]: Yes, exactly. Should have added that, sorry.

**Comment by Hixie** [comment]
@szakarias What is the status of this PR?

**Comment by szakarias** [comment]
I am waiting for review of the corresponding engine PR flutter/engine#4343. cc @chinmaygarde.

**Comment by chinmaygarde** [comment]
I have added a few more comments to the engine PR. I am still a bit unclear about how we are going to address the following issues:

* How are the `*.pbxproj` template upgrades going to be applied to existing users of Flutter? I see no indication about providing upgrade guidance or warning in either the engine or framework patches.
  * My hacky attempt to copy the resources manually in the `xcode_backend.sh build` invocation to the directory specified in `${TARGET_BUILD_DIR}/${WRAPPER_NAME}/flutter_assets` seemed to work. I grabbed those variables from the Xcode build log by clicking the "Show environement variables in build log." checkbox. Maybe we dont need the templates to change.
* Hardcoding a single directory named `flutter_assets` means we can't have multiple Flutter applications in the same application bundle.

**Comment by szakarias** [comment]
Reverted the change in asset retrieval. We still return assets as byte data instead of the asset paths. I am postponing this to a later change. 

With respect to upgrading existing projects, there is this PR #13011. 

With respect to having multiple Flutter applications in the same application bundle, I am not sure how hardcoding `flutter_assets` is worse than the already hardcoded `app.flx`?

**Review by tvolkert — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
BTW, this was checked in when the tree was red. Please avoid checking anything in when the tree is red except for attempts to fix the tree, because it makes it much harder to detect, and later identify, additional failures introduced while the previous failure was ongoing. Thanks!

**Comment by mravn-google** [comment]
@Hixie Tree was red because of a test we had identified, but not marked, as flaky. It was so marked a little later (https://github.com/flutter/flutter/pull/13542).

**Comment by Hixie** [comment]
Ah, excellent. It's good to leave a message on the PR when that happens so that everyone knows it was intentional. Thanks for the update!

**Pull request merged**
